[drm] Fix DRM mode setting to use SFML window dimensions

mickelson opened the pull request:
<!--
Thanks a lot for making a contribution to SFML! 🙂

Before you create the pull request, we ask you to check the follow boxes. (For small changes not everything needs to ticked, but the more the better!)

-   [ ] Has this change been discussed on [the forum](https://en.sfml-dev.org/forums/index.php#c3) or in an issue before?
-   [ ] Does the code follow the SFML [Code Style Guide](https://www.sfml-dev.org/style.php)?
-   [ ] Have you provided some example/test code for your changes?
-   [ ] If you have additional steps which need to be performed list them as tasks!
-->

## Description
[drm] Fix DRM mode setting to use SFML window dimensions

- actually use the dimensions of the SFML window as the DRM mode dimensions when setting a DRM mode.
- If SFML window dimensions don't match a valid DRM mode, show the window contents on screen in the current mode instead of simply failing to show anything.
- With these changes the SFML examples are now working for me

This PR is related to the issue #2946

## Tasks

-   [x] Tested on Linux
-   [ ] Tested on Windows
-   [ ] Tested on macOS
-   [ ] Tested on iOS
-   [ ] Tested on Android

## How to test this PR?

Build SFML with the following cmake variables: SFML_USE_DRM=1 SFML_BUILD_EXAMPLES=1

without x running, run the "island" example

eXpl0it3r commented:
Is there a missing include? The CI is complaining with:

```
/home/runner/work/SFML/SFML/src/SFML/Window/DRM/DRMContext.cpp: In function ‘int {anonymous}::initDrm()’:
/home/runner/work/SFML/SFML/src/SFML/Window/DRM/DRMContext.cpp:367:37: error: ‘drmModeGetConnectorTypeName’ was not declared in this scope; did you mean ‘drmModeGetConnectorCurrent’?
  367 |                 const char *tnStr = drmModeGetConnectorTypeName( connector->connector_type );
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                     drmModeGetConnectorCurrent
```

binary1248 commented:
@Omnomios might be interested in this.

Omnomios commented:
I'll be able to check it out in a day or two. Thanks for tagging me in @binary1248 .

mickelson commented:
Hi, It looks likedrmModeGetConnectorTypeName was first added to libdrm about 2 years ago: https://gitlab.freedesktop.org/mesa/drm/-/commit/50f8d517733d24fce6693ffae552f9833e2e6aa9

I can look into adding a check for this

> On Jun 26, 2024, at 11:38 PM, Lukas Dürrenberger ***@***.***> wrote:
> 
> drmModeGetConnectorTypeName

ChrisThrasher commented:
You mention 3 bullet points you're addressing in this one PR. Is it reasonable to break this PR apartment into multiple smaller PRs that address each bullet point individually or would that be excessively difficult? That might help me understand what all is going on. As it stands now I can't tell what changes in this PR are related to which of the 3 bullets.

mickelson commented:
I've updated this PR to just address the DRM mode initialization fix.

ChrisThrasher commented on src/SFML/Window/DRM/DRMContext.cpp:
```diff
@@ -297,11 +297,13 @@ namespace
             if (fileDescriptor < 0)
                 continue;
             result = getResources(fileDescriptor, resources);
+            if(!result && hasMonitorConnected(fileDescriptor, *resources) != 0)
```
```suggestion
            if (!result && hasMonitorConnected(fileDescriptor, *resources) != 0)
```

ChrisThrasher commented on src/SFML/Window/DRM/DRMContext.cpp:
```diff
@@ -394,56 +381,56 @@ namespace
     {
-        if (initialized)
+	// don't do anything if supplied width and height are 0 and we already have a drm mode
+	if (( width == 0 ) && drmNode.mode )
```
```suggestion
        if (( width == 0 ) && drmNode.mode )
```
Please use spaces to indent instead of tabs.

ChrisThrasher commented on src/SFML/Window/DRM/DRMContext.cpp:
```diff
@@ -456,30 +443,41 @@ namespace
-                    deviceString,     // device
-                    modeString,       // requested mode
-                    refreshRate) < 0) // screen refresh rate
+        bool matched=false;
```
```suggestion
        bool matched = false;
```

ChrisThrasher commented on src/SFML/Window/DRM/DRMContext.cpp:
```diff
@@ -522,12 +520,14 @@ m_config     (NULL),
 m_scanOut    (false)
 {
     contextCount++;
+    if ( initDrm() < 0 )
```
```suggestion
    if (initDrm() < 0)
```

ChrisThrasher commented on src/SFML/Window/DRM/DRMContext.cpp:
```diff
@@ -522,12 +520,14 @@ m_config     (NULL),
 {
     contextCount++;
+    if ( initDrm() < 0 )
+	 return;
```
```suggestion
        return;
```

ChrisThrasher commented on src/SFML/Window/DRM/DRMContext.cpp:
```diff
@@ -555,13 +553,20 @@ m_config     (NULL),
 {
     contextCount++;
 
+    if ( initDrm() < 0 )
```
```suggestion
    if (initDrm() < 0)
```

ChrisThrasher commented on src/SFML/Window/DRM/DRMContext.cpp:
```diff
@@ -555,13 +553,20 @@ m_config     (NULL),
+    if (owner)
+        size = owner->getSize();
+
+    setDrmMode( size.x, size.y );
```
```suggestion
    setDrmMode(size.x, size.y);
```

ChrisThrasher commented on src/SFML/Window/DRM/DRMContext.cpp:
```diff
@@ -573,28 +578,27 @@ m_scanOut    (false)
 m_scanOut    (false)
 {
     contextCount++;
+    if ( initDrm() < 0 )
```
```suggestion
    if (initDrm() < 0)
```

ChrisThrasher commented on src/SFML/Window/DRM/DRMContext.cpp:
```diff
@@ -706,7 +710,7 @@ void DRMContext::display()
         {
             err() << "Failed to set mode: " << std::strerror(errno) << std::endl;
-            std::abort();
+            std::exit( EXIT_FAILURE );
```
```suggestion
            std::exit(EXIT_FAILURE);
```

ChrisThrasher commented on src/SFML/Window/DRM/DRMContext.cpp:
```diff
@@ -573,28 +578,27 @@ m_scanOut    (false)
 
 ////////////////////////////////////////////////////////////
-DRMContext::DRMContext(DRMContext* shared, const ContextSettings& settings, unsigned int width, unsigned int height) :
+DRMContext::DRMContext(DRMContext* shared, const ContextSettings& settings, unsigned int /*width*/, unsigned int /*height*/) :
```
Why comment out these parameters instead of just removing them?

mickelson replied:
The parameters need to stay as this constructor is used in Window/GlContext.cpp

ChrisThrasher reviewed (changes requested):
Please take a closer look at the style of the files you're changing and make sure you code matches that style.

eXpl0it3r commented:
SFMLBuildMaster: Build this please.

mickelson commented:
Hi Chris, I think I've addressed all of your comments now.  Let me know if there is anything else I can do to facilitate review of this, I have a few more drm-related PRs I'd like to submit that depend on this one.

ChrisThrasher commented:
I'm unsure if I want to approve this PR for v2 (`2.6.x` branch) as opposed to v3 (`master` branch). A lot about the DRM code has been cleaned up between v2 and v3 so a change as big as this (in addition to future changes you have planned) may cause nasty merge conflicts when these commits are merged into `master` sometime in the future.

mickelson commented:
Hi Chris, if these changes go into 2.6.x I’m happy to do the PR to merge them over to master as well (if that is any help)

Andrew

> On Jul 27, 2024, at 4:12 PM, Chris Thrasher ***@***.***> wrote:
> 
> ﻿
> I'm unsure if I want to approve this PR for v2 (2.6.x branch) as opposed to v3 (master branch). A lot about the DRM code has been cleaned up between v2 and v3 so a change as big as this (in addition to future changes you have planned) may cause nasty merge conflicts when these commits are merged into master sometime in the future.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.

eXpl0it3r commented:
I don't own a RPi 5, is there a way to test this in a VM or WSL2?

We're close to releasing 2.6.2 and it would be great to have this merged & tested before that.

eXpl0it3r reviewed (approved):
Even though I'm unable to test this and it doesn't seem to fix the linked issue for everyone, I believe we can merge this and take it into SFML 2.6.2

eXpl0it3r commented:
SFMLBuildMaster: Build this please.
